src/main/kotlin/com/openlattice/organizations/ExternalDatabaseManagementService.kt

```diff
@@ -0,0 +1,369 @@
+import com.openlattice.postgres.ResultSetAdapters.organizationId
+import com.openlattice.privileges.OwnerPrivileges
+import com.zaxxer.hikari.HikariDataSource
+import org.apache.olingo.commons.api.edm.FullQualifiedName
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Service
+import java.security.Permissions
+import java.time.OffsetDateTime
+import java.util.*
+import kotlin.collections.LinkedHashMap
+
+@Service
+class ExternalDatabaseManagementService(
+        private val hazelcastInstance: HazelcastInstance,
+        private val hds: HikariDataSource,
+        private val assemblerConnectionManager: AssemblerConnectionManager, //for now using this, may need to move connection logic to its own file
+        private val securePrincipalsManager: SecurePrincipalsManager,
+        private val aclKeyReservations: HazelcastAclKeyReservationService,
+        private val authorizationManager: AuthorizationManager
+) {
+
+    private val organizationExternalDatabaseColumns: IMap<UUID, OrganizationExternalDatabaseColumn> = hazelcastInstance.getMap(HazelcastMap.ORGANIZATION_EXTERNAL_DATABASE_COlUMN.name)
+    private val organizationExternalDatabaseTables: IMap<UUID, OrganizationExternalDatabaseTable> = hazelcastInstance.getMap(HazelcastMap.ORGANIZATION_EXTERNAL_DATABASE_TABLE.name)
+    private val securableObjectTypes: IMap<AclKey, SecurableObjectType> = hazelcastInstance.getMap(HazelcastMap.SECURABLE_OBJECT_TYPES.name)
+    private val aces: IMap<AceKey, AceValue> = hazelcastInstance.getMap(HazelcastMap.PERMISSIONS.name)
+    private val logger = LoggerFactory.getLogger(ExternalDatabaseManagementService::class.java)
+
+    /*CREATE*/
+    fun createOrganizationExternalDatabaseTable(orgId: UUID, table: OrganizationExternalDatabaseTable): UUID {
+        val principal = Principals.getCurrentUser()
+        Principals.ensureUser(principal)
+
+        val tableFQN = FullQualifiedName(orgId.toString(), table.name)
+        checkState(organizationExternalDatabaseTables.putIfAbsent(table.id, table) == null,
+                "OrganizationExternalDatabaseTable ${tableFQN.fullQualifiedNameAsString} already exists")
+        aclKeyReservations.reserveIdAndValidateType(table, tableFQN::getFullQualifiedNameAsString)
+
+        val tableAclKey = AclKey(orgId, table.id)
+        authorizationManager.setSecurableObjectType(tableAclKey, SecurableObjectType.OrganizationExternalDatabaseTable)
+        authorizationManager.addPermission(tableAclKey, principal, EnumSet.allOf(Permission::class.java))
+        //eventBus?
+
+        return table.id
+    }
+
+    fun createOrganizationExternalDatabaseColumn(orgId: UUID, column: OrganizationExternalDatabaseColumn): UUID {
+        val principal = Principals.getCurrentUser()
+        Principals.ensureUser(principal)
+
+        checkState(organizationExternalDatabaseTables[column.tableId] != null,
+                "OrganizationExternalDatabaseColumn ${column.name} belongs to a table that does not exist")
```

Comment:
Minor thing: I would also log the table name here

Reply:
Oh good point!